Objects/Converters/ConverterRevit/ConverterRevitShared/Partial Classes/ConvertMaterialQuantities.cs

```diff
@@ -11,69 +10,148 @@ namespace Objects.Converter.Revit
 {
   public partial class ConverterRevit
   {
-    #region MaterialQuantity
     /// <summary>
-    /// Gets the quantitiy of a material in one element
+    /// Material Quantities in Revit are stored in different ways and therefore need to be retrieved 
+    /// using different methods. According to this forum post https://forums.autodesk.com/t5/revit-api-forum/method-getmaterialarea-appears-to-use-different-formulas-for/td-p/11988215
+    /// "Hosts" (whatever that means) will return the area of a single side of the object while other
+    /// objects will return the combined area of every side of the element. Certain MEP element materials 
+    /// are attached to the MEP system that the element belongs to.
     /// </summary>
     /// <param name="element"></param>
-    /// <param name="material"></param>
+    /// <param name="units"></param>
     /// <returns></returns>
-    public Objects.Other.MaterialQuantity MaterialQuantityToSpeckle(DB.Element element, DB.Material material, string units)
+    public IEnumerable<Other.MaterialQuantity> MaterialQuantitiesToSpeckle(DB.Element element, string units)
     {
-      if (material == null || element == null) 
+      if (MaterialAreaAPICallWillReportSingleFace(element))
+      {
+        return GetMaterialQuantitiesFromAPICall(element, units);
+      }
+      else if (MaterialIsAttachedToMEPSystem(element))
+      {
+        MaterialQuantity quantity = GetMaterialQuantityForMEPElement(element, units);
+        return quantity == null ? new List<MaterialQuantity>() : new List<MaterialQuantity>() { quantity };
+      }
+      else
+      {
+        return GetMaterialQuantitiesFromSolids(element, units);
+      }
+    }
+
+    private IEnumerable<MaterialQuantity> GetMaterialQuantitiesFromAPICall(DB.Element element, string units)
+    {
+      foreach (ElementId matId in element.GetMaterialIds(false))
+      {
+        double volume = element.GetMaterialVolume(matId);
+        double area = element.GetMaterialArea(matId, false);
+        yield return Create(element, matId, area, volume, units);
+      }
+    }
+
+    private MaterialQuantity? GetMaterialQuantityForMEPElement(DB.Element element, string units)
+    {
+      DB.Material material = GetMEPSystemRevitMaterial(element);
+      if (material == null)
+      {
         return null;
+      }
 
-      // To-Do: These methods from the Revit API appear to have bugs.
-      double volume = element.GetMaterialVolume(material.Id);
-      double area = element.GetMaterialArea(material.Id, false);
+      DB.Options options = new() { DetailLevel = ViewDetailLevel.Fine };
+      var (solids, _) = GetSolidsAndMeshesFromElement(element, options);
 
-      // Convert revit interal units to speckle commit units
-      double factor = ScaleToSpeckle(1);
-      volume *= factor * factor * factor;
-      area *= factor * factor;
+      var (area, volume) = GetAreaAndVolumeFromSolids(solids);
```

Comment:
Given the types for area volume are not explicit, we should be.

Reply:
👍